internal/core/src/indexbuilder/KmeansMajorCompaction.cpp

```diff
@@ -0,0 +1,351 @@
+        if (selected_size < train_size &&
+            selected_size + file_sizes[idx[i]] >= train_size) {
+            auto cur_size = train_size - selected_size;
+            local_chunk_manager->Read(
+                file_paths[idx[i]],
+                0,
+                reinterpret_cast<char*>(buf.get()) + offset,
+                cur_size);
+            break;
+        } else {
+            selected_size += file_sizes[idx[i]];
+            local_chunk_manager->Read(
+                file_paths[idx[i]],
+                0,
+                reinterpret_cast<char*>(buf.get()) + offset,
+                file_sizes[idx[i]]);
+            offset += file_sizes[idx[i]];
+        }
+    }
+    return buf;
+}
+
+template <typename T>
+BinarySet
+KmeansMajorCompaction<T>::Upload() {
+    BinarySet ret;
+
+    std::unordered_map<std::string, int64_t> remote_paths_to_size;
+    file_manager_->AddCompactionResultFiles(result_files_,
+                                            remote_paths_to_size);
+    for (auto& file : remote_paths_to_size) {
+        ret.Append(file.first, nullptr, file.second);
+    }
+
+    return ret;
+}
+
+void
+WritePBFile(google::protobuf::Message& message, std::string& file_path) {
+    std::ofstream outfile;
+    outfile.open(file_path.data(), std::ios_base::out | std::ios_base::binary);
+    if (outfile.fail()) {
+        std::stringstream err_msg;
+        err_msg << "Error: open local file '" << file_path << " failed, "
+                << strerror(errno);
+        throw SegcoreError(FileOpenFailed, err_msg.str());
+    }
+
+    outfile.seekp(0, std::ios::beg);
+    if (!message.SerializeToOstream(&outfile)) {
+        std::stringstream err_msg;
+        err_msg << "Error: write local file '" << file_path << " failed, "
+                << strerror(errno);
+        throw SegcoreError(FileWriteFailed, err_msg.str());
+    }
+}
+
+template <typename T>
+void
+KmeansMajorCompaction<T>::Train() {
+    if constexpr (!std::is_same_v<T, float>) {
+        PanicInfo(
+            ErrorCode::UnexpectedError,
+            fmt::format("kmeans major compaction only supports float32 now"));
+    }
+    auto insert_files = milvus::index::GetValueFromConfig<
+        std::map<int64_t, std::vector<std::string>>>(config_, "insert_files");
+    AssertInfo(insert_files.has_value(),
+               "insert file paths is empty when major compaction");
+    auto segment_size =
+        milvus::index::GetValueFromConfig<int64_t>(config_, "segment_size");
+    AssertInfo(segment_size.has_value(),
+               "segment size is empty when major compaction");
+    auto train_size =
+        milvus::index::GetValueFromConfig<int64_t>(config_, "train_size");
+    AssertInfo(train_size.has_value(),
+               "train size is empty when major compaction");
+
+    std::vector<std::string> data_files;
+    std::vector<uint64_t> offsets;
+    uint32_t dim = 0;
+    auto data_size = file_manager_->CacheCompactionRawDataToDisk(
```

Comment:
when did cached file cleaned?

We need to check everywhere about the cache file. It should be cleaned if the analysis cleaned or failed